common/kvstore/src/main/java/org/apache/spark/util/kvstore/LevelDB.java

```diff
@@ -322,26 +323,15 @@ public void close() throws IOException {
     }
   }
 
-  /**
-   * Closes the given iterator if the DB is still open. Trying to close a JNI LevelDB handle
-   * with a closed DB can cause JVM crashes, so this ensures that situation does not happen.
-   */
-  void closeIterator(LevelDBIterator<?> it) throws IOException {
-    notifyIteratorClosed(it);
-    synchronized (this._db) {
-      DB _db = this._db.get();
-      if (_db != null) {
-        it.close();
-      }
-    }
-  }
```

Comment:
Any particular reason to move this into `ResourceCleaner.run` ?

Reply:
Good point, this way we can reduce some code changes.

Reply:
What is your opinion on the comments here? @zhaomin1423

Reply:
It was originally used in finalize method, but now it has been moved to ResourceCleaner.run. I think this logic is consistent

Reply:
The logic is consistent - and would remain the same with and without this move - I was wondering why the change ? Whether this was simply refactoring or there is any other reason for it.

Reply:
resource cleaner has a reference to both `dbIterator` and `levelDB` - not sure what I am missing ?

Reply:
Are you worried about the impact of ResourceCleaner on them being cleaned up by the JVM

Reply:
I am trying to understand why this needs to be moved, was there any specific reason to do so ?
If no, why not revert it back ?

Reply:
My understanding is that it is necessary to hold their references, and I have not yet thought of any other way to convey them

Reply:
We can revert change of `closeIterator` method in `LevelDB` and refactor it as :

```java
  void closeIterator(DBIterator it) throws IOException {
    notifyIteratorClosed(it);
    synchronized (this._db) {
      DB _db = this._db.get();
      if (_db != null) {
        it.close();
      }
    }
  }
```

then there we can call `closeIterator` directly like:

```
public void run() {
      if (started.compareAndSet(true, false)) {
        try {
          levelDB.closeIterator(dbIterator);
        } catch (IOException e) {
          // logXXX("Failed to close iterator", e);
        }
      }
    }
```

Do you mean this idea? @mridulm

Reply:
(I am on vacation, but that sounds right @LuciferYang )

Reply:
Thanks @mridulm 
@zhaomin1423 Is this ok for you?

Reply:
Great, I see.